### PR TITLE
Automatically insert - and * in lists

### LINKED ISSUE #371: Markdown: New list item (* or -) not appearing automatically on new line of a list
When editing a Markdown list in a Android app the editor adds new list item character (\* or -) automatically when hitting enter at the end of a list item.

This doesn't happen on the Windows desktop app.

### PR DESCRIPTION
**Pull request by dmsnell**
Depends on #386 
Resolves #371

Previously, there was a missing behavior in this app that was present in
the other Simplenote apps: when creating a new line after a line
starting with a list bullet, the new line should start with the same
bullet at the same level of indentation.

This PR adds that behavior into the app.

**Testing**
Create a list in a note, such as the following…

```
List:
	- One
	- Two
```

On either of those lines, if you hit _enter_ to create a new line, the tab and dash should come with it as expected. Attempt to indent or outdent those lines. Try to add a second line with _enter, enter_ and the list prefix should disappear, returning the line to normal.

cc: @roundhill @rodrigoi @drw158

### CODE REVIEW

**Comment by nfcampos** [comment]
(this is another one that is better with draft.js, see https://github.com/nfcampos/simplenote-electron/blob/lists/lib/note-content-editor.jsx#L73)

File: lib/note-detail.jsx [hunk 1]
```diff
@@ -1,62 +1,166 @@
...
+		}
+
+		const {
+			selectionStart,
+			value,
+		} = this.noteEditor;
+
+		const prevLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+		const prevLine = value.substring( prevLineStart, selectionStart - 1 );
+		const prevIndent = prevLine.match( /^(\s+)([-*])/ );
```
Comment (nfcampos) [inline lib/note-detail.jsx]: do you mean `[-\*]` instead of `[-*]`?

File: lib/note-detail.jsx [hunk 2]
```diff
@@ -1,62 +1,166 @@
...
+		// if inserting at a cursor position
+		if ( ! event.shiftKey && selectionStart === selectionEnd ) {
+
+			// handle list indents too
+			const currentLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+			const currentLine = value.substring( currentLineStart, selectionStart );
+			console.log( `currentLine: '${ currentLine }'` );
+
+			// line is a list point
+			if ( /^\s+[-*]\s?$/.test( currentLine ) ) {
```
Comment (nfcampos) [inline lib/note-detail.jsx]: it's not an amazing idea to maintain two separate (and similar) regexps for the same thing in each of the two functions, they're bound to get out of sync

File: lib/note-detail.jsx (hunk 1 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: @nfcampos did you find a problem with it? inside of the character list `[]` some of the RegExp symbols take on new meaning, some are even position-dependent. In this case `*` matches a literal `*` and _doesn't mean_ "any number of `-`"

Comment (nfcampos) [inline lib/note-detail.jsx]: ah you're right, nevermind then

File: lib/note-detail.jsx (hunk 2 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: which two are you speaking of? one of the RegExps is for determining if the previous line _has_ a list item while the other is for determining if the previously line _is only_ the list item bullet without following text.

Comment (nfcampos) [inline lib/note-detail.jsx]: `/^(\s+)([-*])([\s\S]+)?/` I mean it's better to use one regexp for both things, if the third group doesn't match then you have "only the list item bullet without following text"

File: lib/note-detail.jsx [hunk 3]
```diff
@@ -1,62 +1,166 @@
...
+
+		// The previous line did not start with
+		// indentation, so continue as normal
+		if ( null === prevIndent ) {
+			return true;
+		}
+
+		const indent = prevIndent[ 1 ];
+		const bullet = prevIndent[ 2 ];
+		const nextIndent = `\n${ indent }${ bullet } `;
```
Comment (nfcampos) [inline lib/note-detail.jsx]: what's the point of `indent` and `bullet` being in separate groups if you then concatenate them?

File: lib/note-detail.jsx (hunk 2 above)
Comment (nfcampos) [inline lib/note-detail.jsx]: (actually it should be `/^(\s+)([-*])\s?([\s\S]+)?/` to allow for a space after the bullet)

File: lib/note-detail.jsx (hunk 3 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: good call. originally I was planning on doing something with these, but now I'm not. I'll take them out

File: lib/note-detail.jsx (hunk 2 above)
Comment (nfcampos) [inline lib/note-detail.jsx]: yeah maybe it's not worth it, i definitely see the point of simpler regexps, i just dont like the possibility of getting bugs from someone changing one of these without changing both

File: lib/note-detail.jsx [hunk 4]
```diff
@@ -1,62 +1,166 @@
...
 		fontSize: PropTypes.number,
 		onChangeContent: PropTypes.func.isRequired
-	},
+	};
+
+	constructor( props ) {
+		super( props );
 
-	componentWillMount: function() {
+		this.storeNoteEditor = r => this.noteEditor = r;
```
Comment (nfcampos) [inline lib/note-detail.jsx]: since you started using class properties these could all be turned into those and the constructor could go away

File: lib/note-detail.jsx [hunk 5]
```diff
@@ -1,62 +1,166 @@
...
-	},
-
-	initializeNoteEditor: function( noteEditor ) {
-		this.noteEditor = noteEditor;
-	},
-
-	isValidNote: function( note ) {
-		return note && note.id;
-	},
+		window.addEventListener( 'keydown', this.interceptTabPresses );
```
Comment (nfcampos) [inline lib/note-detail.jsx]: these can't be regular react event handlers on `Textarea`?

File: lib/note-detail.jsx (hunk 3 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: removed in 3fbfd77

File: lib/note-detail.jsx [hunk 6]
```diff
@@ -1,62 +1,164 @@
...
 
-	componentWillMount: function() {
+		this.storeNoteEditor = r => this.noteEditor = r;
 		this.queueNoteSave = debounce( this.saveNote, saveDelay );
-		this.noteEditor = uninitializedNoteEditor;
-	},
+	}
 
-	componentDidMount: function() {
+	componentDidMount = () => {
```
Comment (nfcampos) [inline lib/note-detail.jsx]: is there a reason lifecycle methods like this one aren't regular methods ie. `componentDidMount() {`? they definitely don't need to be bound to the instance

File: lib/note-detail.jsx (hunk 5 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: @nfcampos the tab is a strange character on the web. the tab key doesn't trigger any of the React keypress events from the textarea because it's designed as a way to advance the input focus across the page. these behaviors we're adding actually break web accessibility in one sense because we're changing what the tab key does. on the other hand, this is more of an app than a webpage and the other apps already hold this behavior.

in short, no we couldn't do it with a React event because it won't fire when we need it

File: lib/note-detail.jsx (hunk 6 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: They would have to be bound because they access `this`, and I prefer this syntax over creating the bind statements in the constructor.

File: lib/note-detail.jsx (hunk 4 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: `queueNoteSave` would still be problematic the way its declared. we could get around this by moving `debounce` to where `saveNote` is defined, but for the time being I want to leave it where it is.

File: lib/note-detail.jsx (hunk 2 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: after playing around with composing the RegExps, I think I want to leave them alone for now. I don't see much gain in combining them and I don't see much gain in composing them when there are just two uses of them and those uses aren't identical and they aren't complicated.

thoughts @rodrigoi ?

File: lib/note-detail.jsx (hunk 5 above)
Comment (nfcampos) [inline lib/note-detail.jsx]: hmm, it does trigger `keydown`, see here a quick example, press tab while focused on that textarea https://esnextb.in/?gist=10f5e3b63353827d220fbe6c4bfe7c05

File: lib/note-detail.jsx (hunk 6 above)
Comment (nfcampos) [inline lib/note-detail.jsx]: nope, you only need to bind class methods to an instance when you use them "detached" from the a reference to the instance, ie.
```js
const obj = {
  abc: 2,

  boo() {
    console.log(this.abc)
  }
}

obj.boo() // logs `2`
// this is akin to how react internally calls lifecycle methods

const boo = this.boo

boo() // fails because `this` is not defined
// because you're not calling this method on any object (and the function isn't previously bound to any context), so there's no context to pass to the function
// this is akin to how event handler end up being called by react, because by doing `<div onClick={this.boo} />` you're passing a reference to the function boo, like in `const boo = this.boo` above
```

long story short, do bind event handlers, don't bind lifecycle methods

File: lib/note-detail.jsx (hunk 5 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: I guess I assumed it was a web thing in general. It's definitely not coming through in the Electron app. Your suggestion was actually my first attempt but then the event never fired. I couldn't seem to easily track down whatever might be intercepting it either.

I'd be more than happy to get rid of the window listener if we can. If you check out this branch and test locally you should be able to see what I'm seeing though.

Comment (nfcampos) [inline lib/note-detail.jsx]: hmm, it works for me https://github.com/nfcampos/simplenote-electron/commit/a0873f388ca3325d556e994cd20aa09cfde93412 running it in chrome, weird

**Comment by nfcampos** [comment]
some stuff isn't right with shift+tab

see this gif http://www.giphy.com/gifs/3o6Zt9ncYZ8oUFjnfa

i first press tab, works as expected, then press shift-tab, and some text I had deleted (I think) gets inserted instead of the tab being removed

File: lib/note-detail.jsx (hunk 5 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: aha, so strange but the `onKeyDown` event sends the tab key while `onKeyPress` skips it. I'm updating the PR. thanks again @nfcampos

Comment (dmsnell) [inline lib/note-detail.jsx]: refactored in f984d70

**Comment by dmsnell** [comment]
> some stuff isn't right with shift+tab

@nfcampos I updated the <kbd>shift+tab</kbd> behavior in 2031dc6. Thanks for finding that!

File: lib/note-detail.jsx (hunk 6 above)
Comment (dmsnell) [inline lib/note-detail.jsx]: TIL. thanks @nfcampos!

refactored in 0d00ddf

File: lib/note-detail.jsx [hunk 7]
```diff
@@ -65,56 +173,78 @@ export default React.createClass( {
...
-		);
-	},
+		document.execCommand( 'insertText', false, indented );
+
+		this.noteEditor.selectionStart = firstLineStart;
+	};
 
-	renderEditable( divStyle ) {
-		const note = this.props.note;
+	render = () => {
```
Comment (nfcampos) [inline lib/note-detail.jsx]: this can also be `render() {` like the other lifecycle methods

File: lib/note-detail.jsx [hunk 8]
```diff
@@ -1,62 +1,170 @@
...
+			value,
+		} = this.noteEditor;
+
+		if ( hasModKeys( [ 'alt', 'ctrl', 'meta', 'shift' ], event ) ) {
+			return true;
+		}
+
+		const prevLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+		const prevLine = value.substring( prevLineStart, selectionStart - 1 );
+		const prevIndent = prevLine.match( /^(\s+[-*])/ );
```
Comment (nfcampos) [inline lib/note-detail.jsx]: since this is a single capturing group with the whole regexp inside it there's no need for it to be a group at all, you can remove the `(` and `)` and use `prevIndent[ 0 ]` below

**Comment by nfcampos** [comment]
another issue, pressing enter with the cursor at the end of line with an indented bullet list item does the wrong thing, see http://www.giphy.com/gifs/3o7TKqHIejvq2zAs36

**Comment by nfcampos** [comment]
also, the behaviour described in the title seems to not be working for lines with bullet items without any indentation, see http://www.giphy.com/gifs/26uf9DOyE3ankSeWc

**Comment by dmsnell** [comment]
> lines with bullet items without any indentation

thanks @nfcampos - we don't consider these lines to be actual lists unless I am wrong, so we wouldn't want a newline to add the `-` or `*`

**Comment by dmsnell** [comment]
> pressing enter with the cursor at the end of line with an indented bullet list item does the wrong thing

good catch! I will make sure that gets corrected before a merge

**Comment by nfcampos** [comment]
> thanks @nfcampos - we don't consider these lines to be actual lists unless I am wrong, so we wouldn't want a newline to add the - or *

@dmsnell ah, i didn't know that. the mac and ios apps do consider those actual lists, i thought the behaviour here would be the same

**Comment by dmsnell** [comment]
> the mac and ios apps do consider those actual lists

haha, another win for reviews. I tested in the Mac app and it cleared it out, but I failed to add anything after the bullet (so of course it wasn't a list when I hit enter).

will update this behavior as well! thanks again

**Comment by drw158** [comment]
I might be doing something wrong, but I couldn't get the new feature to work at all.

Chrome and Electron
Mac OS X

**Comment by dmsnell** [comment]
Closing as irrelevant due to #370
